### ISSUE TITLE
CRASH in os_loader_thread_init_prologue

### ISSUE THREAD

**Issue by trylab** [DynamoRIO/dynamorio#4130]
**Describe the bug**
Crash happens in os_loader_thread_init_prologue

**Screenshots or Pasted Text**
```
2:012> bu dynamorio!os_loader_thread_init_prologue
2:012> g
Breakpoint 0 hit
dynamorio!os_loader_thread_init_prologue:
00000000`710cbd40 48895c2408      mov     qword ptr [rsp+8],rbx ss:00000001`33faf7e0=000001d7aea19080

2:012> p
dynamorio!os_loader_thread_init_prologue+0x5:
00000000`710cbd45 57              push    rdi

2:012> p
dynamorio!os_loader_thread_init_prologue+0x6:
00000000`710cbd46 4883ec20        sub     rsp,20h

2:012> p
dynamorio!os_loader_thread_init_prologue+0xa:
00000000`710cbd4a 488bd9          mov     rbx,rcx

2:012> p
dynamorio!os_loader_thread_init_prologue+0xd:
00000000`710cbd4d 33ff            xor     edi,edi

2:012> p
dynamorio!os_loader_thread_init_prologue+0xf:
00000000`710cbd4f 8b0deb1f0c00    mov     ecx,dword ptr [dynamorio!tls_next_idx (00000000`7118dd40)] ds:00000000`7118dd40=00000001

2:012> p
dynamorio!os_loader_thread_init_prologue+0x15:
00000000`710cbd55 85c9            test    ecx,ecx

2:012> p
dynamorio!os_loader_thread_init_prologue+0x17:
00000000`710cbd57 7e4f            jle     dynamorio!os_loader_thread_init_prologue+0x68 (00000000`710cbda8) [br=0]

2:012> p
dynamorio!os_loader_thread_init_prologue+0x19:
00000000`710cbd59 8b05e51f0c00    mov     eax,dword ptr [dynamorio!tls_array_count (00000000`7118dd44)] ds:00000000`7118dd44=00000000

2:012> p
dynamorio!os_loader_thread_init_prologue+0x1f:
00000000`710cbd5f 85c0            test    eax,eax

2:012> p
dynamorio!os_loader_thread_init_prologue+0x21:
00000000`710cbd61 0f44c1          cmove   eax,ecx

2:012> p
dynamorio!os_loader_thread_init_prologue+0x24:
00000000`710cbd64 488d4fff        lea     rcx,[rdi-1]

2:012> p
dynamorio!os_loader_thread_init_prologue+0x28:
00000000`710cbd68 4863d0          movsxd  rdx,eax

2:012> p
dynamorio!os_loader_thread_init_prologue+0x2b:
00000000`710cbd6b 48c1e203        shl     rdx,3

2:012> p
dynamorio!os_loader_thread_init_prologue+0x2f:
00000000`710cbd6f 8905cf1f0c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118dd44)],eax ds:00000000`7118dd44=00000000

2:012> !address 00000000`7118dd44
Mapping file section regions...
Mapping module regions...
Mapping PEB regions...
Mapping TEB and stack regions...
Mapping heap regions...
Mapping page heap regions...
Mapping other regions...
Mapping stack trace database regions...
Mapping activation context regions...

Usage:                  Image
Base Address:           00000000`7118d000
End Address:            00000000`71199000
Region Size:            00000000`0000c000 (  48.000 kB)
State:                  00001000          MEM_COMMIT
Protect:                00000004          PAGE_READWRITE
Type:                   01000000          MEM_IMAGE
Allocation Base:        00000000`71000000
Allocation Protect:     00000080          PAGE_EXECUTE_WRITECOPY


2:012> g
Breakpoint 0 hit
dynamorio!os_loader_thread_init_prologue:
00000000`710cbd40 48895c2408      mov     qword ptr [rsp+8],rbx ss:00000001`345fed78=000001d7ae9fd9c0

2:011> p
dynamorio!os_loader_thread_init_prologue+0x5:
00000000`710cbd45 57              push    rdi

2:011> p
dynamorio!os_loader_thread_init_prologue+0x6:
00000000`710cbd46 4883ec20        sub     rsp,20h

2:011> p
dynamorio!os_loader_thread_init_prologue+0xa:
00000000`710cbd4a 488bd9          mov     rbx,rcx

2:011> p
dynamorio!os_loader_thread_init_prologue+0xd:
00000000`710cbd4d 33ff            xor     edi,edi

2:011> p
dynamorio!os_loader_thread_init_prologue+0xf:
00000000`710cbd4f 8b0deb1f0c00    mov     ecx,dword ptr [dynamorio!tls_next_idx (00000000`7118dd40)] ds:00000000`7118dd40=00000001

2:011> p
dynamorio!os_loader_thread_init_prologue+0x15:
00000000`710cbd55 85c9            test    ecx,ecx

2:011> p
dynamorio!os_loader_thread_init_prologue+0x17:
00000000`710cbd57 7e4f            jle     dynamorio!os_loader_thread_init_prologue+0x68 (00000000`710cbda8) [br=0]

2:011> p
dynamorio!os_loader_thread_init_prologue+0x19:
00000000`710cbd59 8b05e51f0c00    mov     eax,dword ptr [dynamorio!tls_array_count (00000000`7118dd44)] ds:00000000`7118dd44=00000001

2:011> p
dynamorio!os_loader_thread_init_prologue+0x1f:
00000000`710cbd5f 85c0            test    eax,eax

2:011> p
dynamorio!os_loader_thread_init_prologue+0x21:
00000000`710cbd61 0f44c1          cmove   eax,ecx

2:011> p
dynamorio!os_loader_thread_init_prologue+0x24:
00000000`710cbd64 488d4fff        lea     rcx,[rdi-1]

2:011> p
dynamorio!os_loader_thread_init_prologue+0x28:
00000000`710cbd68 4863d0          movsxd  rdx,eax

2:011> p
dynamorio!os_loader_thread_init_prologue+0x2b:
00000000`710cbd6b 48c1e203        shl     rdx,3

2:011> p
dynamorio!os_loader_thread_init_prologue+0x2f:
00000000`710cbd6f 8905cf1f0c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001

2:011> !address 00000000`7118dd44
Mapping file section regions...
Mapping module regions...
Mapping PEB regions...
Mapping TEB and stack regions...
Mapping heap regions...
Mapping page heap regions...
Mapping other regions...
Mapping stack trace database regions...
Mapping activation context regions...

Usage:                  Image
Base Address:           00000000`7118d000
End Address:            00000000`711a5000
Region Size:            00000000`00018000 (  96.000 kB)
State:                  00001000          MEM_COMMIT
Protect:                00000002          PAGE_READONLY
Type:                   01000000          MEM_IMAGE
Allocation Base:        00000000`71000000
Allocation Protect:     00000080          PAGE_EXECUTE_WRITECOPY


2:011> !heap -p -a 00000000`7118dd44

2:011> p
(11b0.1248): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
dynamorio!os_loader_thread_init_prologue+0x2f:
00000000`710cbd6f 8905cf1f0c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001


2:011> k
 # Child-SP          RetAddr           Call Site
00 00000001`345fed48 00000000`7104472a dynamorio!os_loader_thread_init_prologue+0x2f [c:\projects\dynamorio\core\win32\loader.c @ 384] 
01 00000001`345fed78 00000000`71005615 dynamorio!loader_thread_init+0x2a [c:\projects\dynamorio\core\loader_shared.c @ 277] 
02 00000001`345feda8 00000000`710b4185 dynamorio!dynamo_thread_init+0x515 [c:\projects\dynamorio\core\dynamo.c @ 2394] 
03 00000001`345fee18 00000000`710b3fa5 dynamorio!intercept_new_thread+0xa5 [c:\projects\dynamorio\core\win32\callback.c @ 3139] 
04 00000001`345ff158 00000000`71184ce5 dynamorio!intercept_ldr_init+0x95 [c:\projects\dynamorio\core\win32\callback.c @ 3390] 
05 00000001`345ff188 00000000`00000000 dynamorio!interception_code_array+0xce5
```

**Versions**
 - What version of DynamoRIO are you using? **DynamoRIO-Windows-7.91.18308.zip**
 - Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem? **Not tested**
- What operating system version are you running on? ("Windows 10" is *not* sufficient: give the release number.) **Windows 10 Pro 1909 18363.657**
- Is your application 32-bit or 64-bit? **64-bit**

**Additional context**
Please note ``dynamorio!tls_array_count`` becomes read only when crash happens.

**Comment by derekbruening** [comment]
> Please note `dynamorio!tls_array_count` becomes read only when crash happens.

We make .data read-only after init (https://github.com/DynamoRIO/dynamorio/wiki/Code-Content#security).  However, as you can see in the comment right above the crashing line:

```C
    /* XXX i#4030: We do not support libraries loaded after init time.  We'd have
     * to go through and re-allocate all the thread arrays.
     * We detect this and abort in privload_os_finalize().
     */
```

What client are you running?  What is the command line?  Is your code loading a library with TLS, or is some system library doing that?

**Comment by derekbruening** [comment]
If you could help figure out:

- What library is this with the TLS?  Is it not statically imported at client load time?
- If this library is loaded later, why isn't this detected in privload_os_finalize()?
- If this library is loaded at init, why isn't the TLS set up at init too, well before this new thread is created?

Something is off with the loading of this library.  Either there's a bug in the

**Comment by debach** [comment]
I am observing a similar crash with the precompiled DynamoRIO cronbuild-7.91.18319 on Windows 10 64-bit 1903 (build 18362.535) with a 64-bit client and target. The crash does not happen if I add the `-debug` switch to `drrun.exe`.

I’m not experienced enough to systematically answer your questions (I don’t even know where to start).

What I can say is that I’m suspecting the `Dbghelp.h` library to be the cause of the problems in my client. When my client calls `SymInitialize(GetCurrentProcess(), NULL, true /* or false */)` in the `dr_client_main` function (or anywhere else), it crashes with:

    (3b54.40f4): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    dynamorio!os_loader_thread_init_prologue+0x2f:
    00000000`710cbd9f 89059f1f0c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001
    2:016> k
     # Child-SP          RetAddr           Call Site
    00 000000a2`b57ff228 00000000`7104475a dynamorio!os_loader_thread_init_prologue+0x2f [c:\projects\dynamorio\core\win32\loader.c @ 384] 
    01 000000a2`b57ff258 00000000`71005615 dynamorio!loader_thread_init+0x2a [c:\projects\dynamorio\core\loader_shared.c @ 277] 
    02 000000a2`b57ff288 00000000`710b41b5 dynamorio!dynamo_thread_init+0x515 [c:\projects\dynamorio\core\dynamo.c @ 2419] 
    03 000000a2`b57ff2f8 00000000`710b3fd5 dynamorio!intercept_new_thread+0xa5 [c:\projects\dynamorio\core\win32\callback.c @ 3139] 
    04 000000a2`b57ff638 00000000`71184ce5 dynamorio!intercept_ldr_init+0x95 [c:\projects\dynamorio\core\win32\callback.c @ 3390] 
    05 000000a2`b57ff668 00000000`00000000 dynamorio!interception_code_array+0xce5

Can I do anything about that? (Sorry if this question doesn’t belong here.)

**Comment by derekbruening** [comment]
I think the action item is to figure out what this thread is and when it was created compared to when the TLS is supposed to be initialized.  What is the main thread doing at this point?  Is it not fully initialized?  Is this other thread a very early injected thread and it's racing with the main thread?  Why didn't the os_take_over_all_unknown_threads() take it over?  Is there a race *after* that point?

**Comment by derekbruening** [comment]
And, why is this not happening on other Win10 1909 machines: is there something particular on your machine.

**Comment by debach** [comment]
What I meant is: I don’t have any experience with Windbg and the like, and I had various issues compiling and linking a debug build of DynamoRIO with my test client ;-)

Anyway, I managed to step into my test client with Windbg and narrow down the error a litte. I know this post is lengthy, but I do not mean to hijack this issue – if you feel I should open a new one, let me know.

In the test client, the dbghelp.dll is loaded manually. If instead I use `#pragma comment(lib, "dbghelp.lib")`, the crash happens at `dynamorio!os_loader_thread_init_prologue+0x2f`, as reported by the OP of this issue, before `dr_client_main` is reached.

The test client is:

    #include "dr_api.h"
    #include <DbgHelp.h>

    static void
    event_exit(void);

    DR_EXPORT void
    dr_client_main(client_id_t id, int argc, const char* argv[])
    {
        dr_set_client_name("DynamoRIO Sample Client 'empty'", "http://dynamorio.org/issues");
        dr_register_exit_event(event_exit);
        HMODULE module = LoadLibrary("dbghelp.dll");
    }

    static void
    event_exit(void)
    {
    }

I invoke DynamoRIO with (remember: all is x64):

    windbg.exe drrun.exe -debug -c DynamoRIO_Dbghelp_Test.dll -- calc.exe

In windbg:

    .childdbg 1
    g
    bu DynamoRIO_Dbghelp_Test!dr_client_main
    g
    Breakpoint 0 hit
    DynamoRIO_Dbghelp_Test!dr_client_main
    p
    (we’re at dr_set_client_name)
    p
    (we’re at dr_register_exit_event)
    p
    (we’re at LoadLibrary)
    p

When executing the line with `LoadLibrary`, an error message pops up with:

    Error: DynamoRIO debug check failure:
    dynamorio\core\win32\loader.c:1448
    tls_array_count > opd->tls_idx && tls_array != NULL (Error occurred @0 frags)
    version 7.91.18325, custom build

At that point, the call stack is

     k
     # Child-SP          RetAddr           Call Site
    00 00000071`85bbe038 00000000`713b54af dynamorio!dynamorio_syscall_syscall+0x49 [C:\projects\dynamorio\builds\build-vs17-x64-d9fdabee-debug\core\x86.asm_core.s @ 2518] 
    01 00000071`85bbe040 00000000`713745c6 dynamorio!nt_terminate_process+0x10f [c:\projects\dynamorio\core\win32\ntdll.c @ 2366] 
    02 00000071`85bbe0c0 00000000`7117e732 dynamorio!os_terminate_common+0xd6 [c:\projects\dynamorio\core\win32\os.c @ 1539] 
    03 00000071`85bbe130 00000000`710fd3a2 dynamorio!soft_terminate+0x32 [c:\projects\dynamorio\core\utils.c @ 108] 
    04 00000071`85bbe160 00000000`713fff7b dynamorio!d_r_internal_error+0x1a2 [c:\projects\dynamorio\core\utils.c @ 194] 
    05 00000071`85bbe2e0 00000000`7122e16e dynamorio!privload_call_entry+0x22b [c:\projects\dynamorio\core\win32\loader.c @ 1448] 
    06 00000071`85bbe530 00000000`7122df23 dynamorio!privload_call_entry_if_not_yet+0x9e [c:\projects\dynamorio\core\loader_shared.c @ 124] 
    07 00000071`85bbe570 00000000`7122c945 dynamorio!privload_load_finalize+0xb3 [c:\projects\dynamorio\core\loader_shared.c @ 845] 
    08 00000071`85bbe5e0 00000000`71404cc9 dynamorio!privload_load+0x3f5 [c:\projects\dynamorio\core\loader_shared.c @ 701] 
    09 00000071`85bbe670 00000000`714065cf dynamorio!privload_locate_and_load+0x2e9 [c:\projects\dynamorio\core\win32\loader.c @ 1873] 
    0a 00000071`85bbe700 00000000`714010ee dynamorio!privload_lookup_locate_and_load+0x21f [c:\projects\dynamorio\core\win32\loader.c @ 1142] 
    0b 00000071`85bbe780 00000000`7122ddd6 dynamorio!privload_process_imports+0x32e [c:\projects\dynamorio\core\win32\loader.c @ 1219] 
    0c 00000071`85bbe840 00000000`7122c920 dynamorio!privload_load_process+0xb6 [c:\projects\dynamorio\core\loader_shared.c @ 827] 
    0d 00000071`85bbe890 00000000`71404bb8 dynamorio!privload_load+0x3d0 [c:\projects\dynamorio\core\loader_shared.c @ 695] 
    0e 00000071`85bbe920 00000000`714065cf dynamorio!privload_locate_and_load+0x1d8 [c:\projects\dynamorio\core\win32\loader.c @ 1856] 
    0f 00000071`85bbe9b0 00000000`713ff5ad dynamorio!privload_lookup_locate_and_load+0x21f [c:\projects\dynamorio\core\win32\loader.c @ 1142] 
    10 00000071`85bbea30 00000000`7122bd2c dynamorio!privload_load_private_library+0x3d [c:\projects\dynamorio\core\win32\loader.c @ 1157] 
    11 00000071`85bbea80 00000000`7140cf4f dynamorio!locate_and_load_private_library+0x2c [c:\projects\dynamorio\core\loader_shared.c @ 345] 
    12 00000071`85bbeab0 00000000`7140cc3a dynamorio!helper_LoadLibrary+0x17f [c:\projects\dynamorio\core\win32\drwinapi\kernel32_lib.c @ 146] 
    13 00000071`85bbec00 00007ff6`d386103d dynamorio!redirect_LoadLibraryA+0xba [c:\projects\dynamorio\core\win32\drwinapi\kernel32_lib.c @ 170] 
    14 00000071`85bbec50 00000000`71246df9 DynamoRIO_Dbghelp_Test!dr_client_main+0x3d [c:\projects\dynamorio_dbghelp_test\dynamorio_dbghelp_test.cpp @ 12] 
    15 00000071`85bbec90 00000000`710109d0 dynamorio!instrument_init+0x249 [c:\projects\dynamorio\core\lib\instrument.c @ 770] 
    16 00000071`85bbed30 00000000`7133f9ae dynamorio!dynamorio_app_init+0x6e0 [c:\projects\dynamorio\core\dynamo.c @ 684] 
    17 00000071`85bbf6a0 00000000`71412e40 dynamorio!auto_setup+0x2e [c:\projects\dynamorio\core\arch\x86_code.c @ 169] 
    18 00000071`85bbf7a0 00000071`85bbf7c0 dynamorio!dynamo_auto_start+0x10 [C:\projects\dynamorio\builds\build-vs17-x64-d9fdabee-debug\core\x86.asm_core.s @ 1600] 
    19 00000071`85bbf7a8 00000000`71412e30 0x00000071`85bbf7c0
    1a 00000071`85bbf7b0 00000000`00000000 dynamorio!dynamo_auto_start [C:\projects\dynamorio\builds\build-vs17-x64-d9fdabee-debug\core\x86.asm_core.s @ 1592] 

Switching the stack to 05, where the failing assertion was:

    1:004> dx Debugger.Sessions[0].Processes[21020].Threads[25888].Stack.Frames[5].SwitchTo();dv /t /v
    Debugger.Sessions[0].Processes[21020].Threads[25888].Stack.Frames[5].SwitchTo()
    00000071`85bbe3c8 unsigned char ** tls_array = 0x00000000`00000000
    00000071`85bbe530 struct _dcontext_t * dcontext = 0x000001c1`37929080
    00000071`85bbe538 struct _privmod_t * privmod = 0x000001c1`3793b3d0
    00000071`85bbe540 unsigned int reason = 1
    00000071`85bbe338 struct _os_privmod_data_t * opd = 0x000001c1`3793b5d8
    00000071`85bbe370 unsigned char * entry = 0xffffffff`00000000 "--- memory read error at address 0xffffffff`00000000 ---"
    00000071`85bbe330 char call_routines = 0n1 ''
    00000071`85bbe331 char return_val = 0n1 ''

So, the second part of the failing assertion, `tls_array != NULL`, is false. In fact, the first part of the assertion, `tls_array_count > opd->tls_idx`, is also false:

    1:004> dt tls_array_count
    dynamorio!tls_array_count
    0n0

    1:004> dt opd
    Local var @ 0x7185bbe338 Type _os_privmod_data_t*
    0x000001c1`3793b5d8 
       +0x000 tls_callbacks    : (null) 
       +0x008 tls_callback_count : 0n0
       +0x00c tls_idx          : 0n0
       +0x010 tls_init_data    : 0x000001c1`b79327c8  ""
       +0x018 tls_size         : 8

Any thoughts?

**Comment by debach** [comment]
What I find odd is that in `privload_locate_and_load` (stack frame 0xe) the variable `impname` is `"dbghelp.dll"`, but in `privload_load` (stack frame 0xd) the variable `filename` is `"C:\Windows/system32/msvcrt.dll"`. Does this indicate a race condition on `modpath`?

**Comment by debach** [comment]
One more thing: while loading dbghelp.dll, DynamoRIO tries to load msvcrt.dll in privload_process_imports@1219 (stack frame 0x0b in the above stack). This is the library that (apparently) finally causes the error.

**Comment by derekbruening** [comment]
Thank you, this is helpful.  A couple of comments:

Currently, static TLS is not supported in dynamically loaded libraries (LoadLibrary from dr_client_main).  There is code that is supposed to provide a nice visible fatal error message at https://github.com/DynamoRIO/dynamorio/blob/master/core/win32/loader.c#L2473 but it looks like it is called *after* privload_process_imports(), so you end up hitting the other assert before it gets to the fatal message.  So that message is not doing its job.

However, static TLS should work for the statically imported dbghelp (when you did the pragma).  We have a number of clients who use dbghelp and they have never hit a problem, so I'm not understanding why it is failing.  I wonder if you could get the callstacks for that case?

For the dynamic load, there is an action item to change the code to have the fatal message raised *before* the assert.  Furthermore, if a dynamic load is something people really want to support, and someone was interested in adding such code, we would be happy to take a PR adding that feature.

**Comment by debach** [comment]
That’s good to know, thank you:

> Currently, static TLS is not supported in dynamically loaded libraries

> We have a number of clients who use dbghelp and they have never hit a problem, so I'm not understanding why it is failing.

Going back to `#pragma comment(lib, "dbghelp.lib")`, the following client causes crashes:

    #include "dr_api.h"
    #include <DbgHelp.h>

    #pragma comment(lib, "dbghelp.lib")

    static void
    event_exit(void);

    DR_EXPORT void
    dr_client_main(client_id_t id, int argc, const char* argv[])
    {
        dr_set_client_name("DynamoRIO Sample Client 'empty'", "http://dynamorio.org/issues");
        dr_register_exit_event(event_exit);
        SymInitialize(GetCurrentProcess(), NULL, true);
    }

    static void
    event_exit(void)
    {
    }

The invokation is without the `-debug` flag, tested with two targets:

    windbg.exe drrun.exe -c DynamoRIO_Dbghelp_Test.dll -- calc.exe
    windbg.exe drrun.exe -c DynamoRIO_Dbghelp_Test.dll -- notepad.exe

The stack trace is the same as that of OP for both targets (literally, except RetAddr):

        (25bc.3320): Break instruction exception - code 80000003 (first chance)
    ntdll!LdrpDoDebuggerBreak+0x30:
    00007ff8`471511dc cc              int     3
    1:004> g
    ModLoad: 00007ff8`45ec0000 00007ff8`45eee000   C:\Windows\System32\IMM32.DLL
    ModLoad: 00007ff8`43e00000 00007ff8`43e3c000   C:\Windows\system32\SophosAV\SOPHOS~1.DLL
    ModLoad: 00007ff8`45e00000 00007ff8`45e08000   C:\Windows\System32\PSAPI.DLL
    ModLoad: 00000000`71000000 00000000`711b1000   C:\projects\dynamorio\builds\build-vs17-x64-d9fdabee-relwithdebinfo\lib64\release\dynamorio.dll
    ModLoad: 00000000`72000000 00000000`72053000   DynamoRIO_Dbghelp_Test.dll
    ModLoad: 00007ff6`d3860000 00007ff6`d38b3000   DynamoRIO_Dbghelp_Test.dll
    ModLoad: 00000167`58380000 00000167`58432000   KERNEL32.dll
    ModLoad: 00000167`58510000 00000167`587b3000   KERNELBASE.dll
    ModLoad: 00007fff`f65b0000 00007fff`f6738000   dbghelp.dll
    ModLoad: 00007ff6`d38c0000 00007ff6`d3a48000   dbghelp.dll
    ModLoad: 00000167`58440000 00000167`584de000   msvcrt.dll
    ModLoad: 00000167`587c0000 00000167`58863000   ADVAPI32.dll
    ModLoad: 00000167`58870000 00000167`58907000   SECHOST.dll
    ModLoad: 00000167`58910000 00000167`58a30000   RPCRT4.dll
    (25bc.4e7c): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    dynamorio!os_loader_thread_init_prologue+0x2f:
    00000000`710cb5af 89058f270c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001
    1:008> k
     # Child-SP          RetAddr           Call Site
    00 0000004d`6beff018 00000000`7104480a dynamorio!os_loader_thread_init_prologue+0x2f [c:\projects\projekte\dynamorio\core\win32\loader.c @ 384] 
    01 0000004d`6beff048 00000000`71005635 dynamorio!loader_thread_init+0x2a [c:\projects\projekte\dynamorio\core\loader_shared.c @ 277] 
    02 0000004d`6beff078 00000000`710b39a5 dynamorio!dynamo_thread_init+0x515 [c:\projects\projekte\dynamorio\core\dynamo.c @ 2419] 
    03 0000004d`6beff0e8 00000000`710b37c5 dynamorio!intercept_new_thread+0xa5 [c:\projects\projekte\dynamorio\core\win32\callback.c @ 3139] 
    04 0000004d`6beff428 00000000`71184ce5 dynamorio!intercept_ldr_init+0x95 [c:\projects\projekte\dynamorio\core\win32\callback.c @ 3390] 
    05 0000004d`6beff458 00000000`00000000 dynamorio!interception_code_array+0xce5

Is it normal that the same module is loaded twice (dbghelp and the client module)? It’s a company policy to use Sophos Antivirus (I know…), but I will check if I get any errors on my personal computer at the weekend that doesn’t have Sophos.

With the `-debug` flag it gets a bit weird. Call stack:

    windbg.exe drrun.exe -debug -c DynamoRIO_Dbghelp_Test.dll -- calc.exe

Output:

    (6340.5f38): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    ntdll!LdrpHandleTlsData+0x43f:
    00007ff8`470d5723 4d8b40f8        mov     r8,qword ptr [r8-8] ds:cdcdcdcd`cdcdcdc5=????????????????
    1:008> k
     # Child-SP          RetAddr           Call Site
    00 00000087`b0b7cd70 00007ff8`470d529d ntdll!LdrpHandleTlsData+0x43f
    01 00000087`b0b7ce70 00007ff8`4709cb2a ntdll!LdrpDoPostSnapWork+0x6d
    <many more frames>
    27 00000087`b0b7f860 00007ff8`470eced1 KERNEL32!BaseThreadInitThunk+0x14
    28 00000087`b0b7f890 00000000`00000000 ntdll!RtlUserThreadStart+0x21

Recall this is with the target `calc.exe`. For the target `notepad.exe`, everything works fine at first with `-debug`. However, if I open the “File → Open“ dialogue, I get 

    (6210.4d1c): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    00007ff5`ea305584 4d8b40f8        mov     r8,qword ptr [r8-8] ds:feeefeee`feeefee6=????????????????
    1:014> k
     # Child-SP          RetAddr           Call Site
    00 000000a4`372fc1c0 00000000`00000004 0x00007ff5`ea305584
    01 000000a4`372fc1c8 00000000`00000000 0x4

as well as a DynamoRIO error message about a failed assertion at core\dispatch.c:757.

I tested this with a third target, IrfanView, which is a popular image viewer on Windows. Without the debug switch, I get the same error and back trace as for the other two targets (the one reported by OP).  With the debug switch, when I use the “File → Open” dialogue, I sometimes get this:

    <lots of ModLoads>
    ModLoad: 00007ff8`41ed0000 00007ff8`41ef9000   C:\Windows\system32\RMCLIENT.dll
    Critical error detected c0000374
    (5eac.268c): Break instruction exception - code 80000003 (first chance)
    ntdll!RtlReportCriticalFailure+0x56:
    00007ff8`471791f2 cc              int     3
    1:004> k
     # Child-SP          RetAddr           Call Site
    00 00000042`c7f71c90 00007ff8`47181622 ntdll!RtlReportCriticalFailure+0x56
    01 00000042`c7f71d80 00007ff8`4718192a ntdll!RtlpHeapHandleError+0x12
    02 00000042`c7f71db0 00007ff8`4718a8e9 ntdll!RtlpHpHeapHandleError+0x7a
    03 00000042`c7f71de0 00007ff8`470c07df ntdll!RtlpLogHeapFailure+0x45
    04 00000042`c7f71e10 00007ff8`470bfc11 ntdll!RtlpFreeHeapInternal+0x75f
    05 00000042`c7f71ec0 00007ff8`470d5731 ntdll!RtlFreeHeap+0x51
    06 00000042`c7f71f00 00007ff8`470d529d ntdll!LdrpHandleTlsData+0x44d
    <... 35 more stack frames>

and sometimes:

    <more ModLoads>
    ModLoad: 00007ff8`41ed0000 00007ff8`41ef9000   C:\Windows\system32\RMCLIENT.dll
    (2aa4.12a0): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    00007ff6`845b2378 4d8b40f8        mov     r8,qword ptr [r8-8] ds:abababab`abababa3=????????????????
    1:004> k
     # Child-SP          RetAddr           Call Site
    00 000000a0`070f1d70 00000000`00000004 0x00007ff6`845b2378
    01 000000a0`070f1d78 00000000`00000000 0x4
    1:004> gn
    (4ce4.38bc): Access violation - code c0000005 (!!! second chance !!!)
    ntdll!LdrpHandleTlsData+0x43f:
    00007ff8`470d5723 4d8b40f8        mov     r8,qword ptr [r8-8] ds:abababab`abababa3=????????????????
    1:004> k
     # Child-SP          RetAddr           Call Site
    00 000000c3`f99d1b50 00007ff8`470d529d ntdll!LdrpHandleTlsData+0x43f
    <many more frames>
    gn
    1:004> gn
    dynamorio!opnd_create_base_disp+0x3d:
    00000000`7129433d 89442428        mov     dword ptr [rsp+28h],eax ss:00000203`e6eadce8=e6ab7104

I think I will test this on another computer and and see if there are similar issues.

**Comment by derekbruening** [comment]
What is the main thread doing at the time this new thread hits the access violation in os_loader_thread_init_prologue?  Ideally, what is the win32 start address of the new thread: who created it.

**Comment by debach** [comment]
I did not attempt to debug this issue, but I wanted to say I’m now able to run my client. I switched to DynamoRIO-Windows-7.91.18263-0 (from DynamoRIO-Windows-7.91.18319) and compiled my client with Visual Studio 2013 (instead of VS 2017).

**Comment by derekbruening** [comment]
But which one made it go away?  The newer DR, or using 2013?

**Comment by debach** [comment]
Appologies – I knew you wouldn’t let me get away with this ;-)

I just tested and obtained the following. The first column gives the version of DynamoRIO that I used both when building my client (with DDynamoRIO_DIR) and for drrun. The VS13 and VS17 column indicate which version of Visual Studio I’ve used when building my client.

DR Version / Client Version | VS13 | VS17
----------|-------|--------
7.91.18263-0 | good | good
7.91.18319   | bad | bad

I had quite immediate crashes with 7.91.18319 for both VS versions, and everything worked smooth with 7.91.18263-0 for both VS versions.

**Comment by debach** [comment]
I just did a manual binary search and found that the last release with Windows binaries that works for me is cronbuild-7.91.18278. The next release, cronbuild-7.91.18285, does not have Windows binaries and I didn’t test it. The next release with Windows binaries, cronbuild-7.91.18292, crashes for me. Tested with my custom client built with VS2013.

**Comment by derekbruening** [comment]
The culprit is almost certainly 9293e7ae7709ee78fe520722dbe062b8680126d6 for #4030, but that static TLS support is necessary for multiple use cases with VS2017 and with other libraries that require it, so we can't just revert it: this bug will have to be figured out.  What version of dbghelp.dll are you using?

**Comment by debach** [comment]
In my client source code, I have the line

    #pragma comment( lib, "C:\\Program Files (x86)\\Windows Kits\\10\\Lib\\10.0.18362.0\\um\\x64\\dbghelp.lib" )

For reference, the complete client code is:


    #include "dr_api.h"
    #include <DbgHelp.h>

    #pragma comment( lib, "C:\\Program Files (x86)\\Windows Kits\\10\\Lib\\10.0.18362.0\\um\\x64\\dbghelp.lib" )

    static void
    event_exit(void);

    DR_EXPORT void
    dr_client_main(client_id_t id, int argc, const char* argv[])
    {
        /* empty client */
        dr_set_client_name("DynamoRIO Sample Client 'empty'", "http://dynamorio.org/issues");
        dr_register_exit_event(event_exit);
        
        SymInitialize(GetCurrentProcess(), NULL, false);
    }

    static void
    event_exit(void)
    {
        /* empty client */
    }

I compiled with VS2013 against 7.91.18292 with RelWithDebInfo configuration and ran `drrun.exe -c DynamoRIO_Dbghelp_Test.dll -- notepad.exe`. In WinDbg64, I find via `lm -v`

    00007ff7`78380000 00007ff7`7850b000   dbghelp    (deferred)             
        Image path: dbghelp.dll
        Image name: dbghelp.dll
        Browse all global symbols  functions  data
        Timestamp:        Wed Jun 17 07:25:05 2015 (558104B1)
        CheckSum:         0017600C
        ImageSize:        0018B000
        File version:     10.0.10150.0
        Product version:  10.0.10150.0
        File flags:       0 (Mask 3F)
        File OS:          40004 NT Win32
        File type:        2.0 Dll
        File date:        00000000.00000000
        Translations:     0409.04b0
        Information from resource tables:
            CompanyName:      Microsoft Corporation
            ProductName:      Microsoft® Windows® Operating System
            InternalName:     DBGHELP.DLL
            OriginalFilename: DBGHELP.DLL
            ProductVersion:   10.0.10150.0
            FileVersion:      10.0.10150.0 (debuggers(dbg).150616-1659)
            FileDescription:  Windows Image Helper
            LegalCopyright:   © Microsoft Corporation. All rights reserved.

Everything seems to work until I hit my client’s main. Proceeding with `g` gives me:

    (189c.3340): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    dynamorio!dr_symbol_export_iterator_stop+0x68bf:
    00000000`710cbd1f 89051f200c00    mov     dword ptr [dynamorio!unique_build_number+0x32b54 (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001
    1:006> k
     # Child-SP          RetAddr           Call Site
    00 0000004e`b68ff3a8 00007ff8`cad34060 dynamorio!unique_build_number+0x2a242
    01 0000004e`b68ff3b0 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000004e`b68ff770 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000004e`b68ff7a0 00000000`00000000 ntdll!RtlUserThreadStart+0x21

I single-stepped from my client’s `dr_client_main` to the crash. Here’s what happens in compressed form:

    DynamoRIO_Dbghelp_Test!dr_client_main:
    ...
    DynamoRIO_Dbghelp_Test!dr_client_main+0x35:
    dbghelp!SymInitialize:
    ...
    dbghelp!SymInitialize+0x49:
    dynamorio!instrlist_meta_fault_preinsert+0xdea:
    ...
    dynamorio!instrlist_meta_fault_preinsert+0xff3:
    dynamorio!dynamorio_app_init+0x234:
    ...
    dynamorio!dynamorio_app_init+0x390:
    dynamorio!proc_save_fpstate+0xc9fc:
    ...
    dynamorio!proc_save_fpstate+0xcb2d:
    ModLoad: 00007ff8`c7bd0000 00007ff8`c7be1000   C:\Windows\System32\kernel.appcore.dll
    ModLoad: 00007ff8`ca750000 00007ff8`ca7f2000   C:\Windows\System32\clbcatq.dll
    ModLoad: 00007ff8`b71e0000 00007ff8`b72f2000   C:\Windows\System32\MrmCoreR.dll
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(665)\MrmCoreR.dll!00007FF8B720A902: (caller: 00007FF8B71F4732) ReturnHr(1) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(558)\MrmCoreR.dll!00007FF8B71F4875: (caller: 00007FF8B71F24F1) ReturnHr(2) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(296)\MrmCoreR.dll!00007FF8B71F25E0: (caller: 00007FF8B71E75FE) ReturnHr(3) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(114)\MrmCoreR.dll!00007FF8B71E763D: (caller: 00007FF8B71E754A) ReturnHr(4) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(1377)\MrmCoreR.dll!00007FF8B71E75D1: (caller: 00007FF8B71E7446) ReturnHr(5) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\src\cresourcemanagerinternal.cpp(1281)\MrmCoreR.dll!00007FF8B71E7476: (caller: 00007FF8B71E91A3) ReturnHr(6) tid(4498) 80070002 The system cannot find the file specified.
    onecoreuap\base\mrt\runtime\com\winrt\core\lib\winrtresourcemanager.cpp(601)\MrmCoreR.dll!00007FF8B71E9204: (caller: 00007FF7D832FDF6) ReturnHr(7) tid(4498) 80070002 The system cannot find the file specified.
    ModLoad: 00007ff8`ca3b0000 00007ff8`ca4e6000   C:\Windows\System32\MSCTF.dll
    ModLoad: 00007ff8`cabf0000 00007ff8`cacb4000   C:\Windows\System32\OLEAUT32.dll
    ModLoad: 00007ff8`c7ef0000 00007ff8`c8670000   C:\Windows\System32\windows.storage.dll
    ModLoad: 00007ff8`c7c60000 00007ff8`c7c83000   C:\Windows\System32\profapi.dll
    ModLoad: 00007ff8`c7bf0000 00007ff8`c7c3a000   C:\Windows\System32\powrprof.dll
    ModLoad: 00007ff8`c7bc0000 00007ff8`c7bd0000   C:\Windows\System32\UMPDC.dll
    ModLoad: 00007ff8`ca8b0000 00007ff8`ca902000   C:\Windows\System32\shlwapi.dll
    ModLoad: 00007ff8`88320000 00007ff8`883f7000   C:\Windows\System32\efswrt.dll
    ModLoad: 00007ff8`aa290000 00007ff8`aa2ab000   C:\Windows\System32\MPR.dll
    ModLoad: 00007ff8`c3210000 00007ff8`c3363000   C:\Windows\SYSTEM32\wintypes.dll
    ModLoad: 00007ff8`c56f0000 00007ff8`c594a000   C:\Windows\System32\twinapi.appcore.dll
    ModLoad: 00007ff8`c5b10000 00007ff8`c5b39000   C:\Windows\System32\RMCLIENT.dll
    (4518.4054): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    dynamorio!dr_symbol_export_iterator_stop+0x68bf:
    00000000`710cbd1f 89051f200c00    mov     dword ptr [dynamorio!unique_build_number+0x32b54 (00000000`7118dd44)],eax ds:00000000`7118dd44=00000001

**Comment by debach** [comment]
Quick update: the last single-step I did was

    dynamorio!proc_save_fpstate+0xcb2d:
    00000000`7109be9d e8a8030000      call    dynamorio!dynamo_auto_start+0x1a (00000000`7109c24a)
    1:004> p

**Comment by derekbruening** [comment]
It looks like you don't have symbols for dynamorio.dll: those offsets are too large and only involve public symbols (e.g., this is an enormous offset: `dynamorio!unique_build_number+0x2a242`).  https://github.com/DynamoRIO/dynamorio/wiki/Debugging shows the `load_syms64` script which can be used after DR is initialized; before it's initialized windbg should see the library on its own, but maybe it just needs its `.sympath` fixed up.

**Comment by debach** [comment]
I fixed the symbols. I’m now using a custom RelWithDebInfo build of DynamoRIO 7.91.18292 compiled by VS2013 for Win64. To come back to an earlier question:

The call stack of the main thread at the time of the crash is:

    1:008> ~4k
     # Child-SP          RetAddr           Call Site
    00 000000ff`a657dd38 00007ff8`c8b38bc3 ntdll!NtWaitForSingleObject+0x14
    01 000000ff`a657dd40 00007ff8`c5729ee3 KERNELBASE!WaitForSingleObjectEx+0x93
    02 000000ff`a657dde0 00007ff8`c5729c99 twinapi_appcore!Windows::ApplicationModel::Core::CoreApplication::CreateInternal+0x233
    03 000000ff`a657e230 00007ff8`cad0b05f twinapi_appcore!Windows::ApplicationModel::Core::CoreApplication::Create+0x9
    04 000000ff`a657e260 00007ff8`c8b6217b ntdll!RtlRunOnceExecuteOnce+0x8f
    05 000000ff`a657e2a0 00007ff8`c57371f2 KERNELBASE!InitOnceExecuteOnce+0xb
    06 000000ff`a657e2d0 00007ff8`c5740928 twinapi_appcore!Windows::ApplicationModel::Core::CoreApplication::get_CurrentInternal+0x32
    07 000000ff`a657e370 00007ff8`883581c9 twinapi_appcore!Windows::ApplicationModel::Core::CoreApplicationFactory::get_PrivateProperties+0x28
    08 000000ff`a657e3a0 00007ff8`88359fd4 efswrt!Windows::Security::EnterpriseData::ProtectionPolicyManagerStaticsPrivate::GetEventsSingleton+0xd9
    09 000000ff`a657e4b0 00007ff8`c7f9d9fa efswrt!Windows::Security::EnterpriseData::ProtectionPolicyManagerStaticsPrivate::add_PolicyChanged+0x34
    0a 000000ff`a657e500 00007ff7`cfffeb81 windows_storage!Windows::Security::EnterpriseData::ProtectionPolicyManagerStatics::add_PolicyChanged+0x6a
    0b 000000ff`a657e560 00007ff7`cfff06c1 notepad!ProtectionPolicyManager_RegisterForProtectionPolicyChanged+0x129
    0c 000000ff`a657e5d0 00007ff7`cffea306 notepad!NPInit+0x1e5
    0d 000000ff`a657f8d0 00007ff7`d00002c7 notepad!WinMain+0x1b6
    0e 000000ff`a657f9a0 00007ff8`ca9c7bd4 notepad!__mainCRTStartup+0x19f
    0f 000000ff`a657fa60 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    10 000000ff`a657fa90 00000000`00000000 ntdll!RtlUserThreadStart+0x21

The call stack of the crashing thread is that of the OP:

    1:008> ~8
    .  8  Id: 11a0.4864 Suspend: 1 Teb: 000000a1`473f3000 Unfrozen
          Start: ntdll!TppWorkerThread (00007ff8`cad33d60)
          Priority: 0  Priority class: 32  Affinity: ff
    1:008> ~8k
     # Child-SP          RetAddr           Call Site
    00 000000ff`a697ef28 00000000`71043bda dynamorio!os_loader_thread_init_prologue+0x2b [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 385] 
    01 000000ff`a697ef58 00000000`710055a9 dynamorio!loader_thread_init+0x2a [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 277] 
    02 000000ff`a697ef88 00000000`710b3aa0 dynamorio!dynamo_thread_init+0x289 [c:\users\myself\projects\dynamorio\core\dynamo.c @ 2419] 
    03 000000ff`a697efc8 00000000`710b38b5 dynamorio!intercept_new_thread+0xb0 [c:\users\myself\projects\dynamorio\core\win32\callback.c @ 3139] 
    04 000000ff`a697f308 00000000`71185ce5 dynamorio!intercept_ldr_init+0x95 [c:\users\myself\projects\dynamorio\core\win32\callback.c @ 3390] 
    05 000000ff`a697f338 00000000`00000000 dynamorio!interception_code_array+0xce5

I’ve been unable so far to figure out why the new thread, i.e., the crashing thread ~8, is created and who creates it.

**Comment by derekbruening** [comment]
My theory is that this has nothing to do with dbghelp.dll, that your version of dbghelp.dll has no static TLS, and that there's some other library being dynamically loaded.

If you could put a breakpoint here:
https://github.com/DynamoRIO/dynamorio/blob/master/core/win32/loader.c#L2469
and print out the library (`mod_name`) and callstack for each occurrence?

**Comment by debach** [comment]
The only module at that breakpoint *seems* (see below) to be `C:\Windows/system32/msvcrt.dll`, according to the global `modpath`. I did the following:

    0:000> .childdbg 1
    0:000> g
    (51f0.49d8): Break instruction exception - code 80000003 (first chance)
    ntdll!LdrpDoDebuggerBreak+0x30:
    00007ff8`cadd121c cc              int     3
    1:004> bu dynamorio!privload_os_finalize+0x110
    1:004> g
    Breakpoint 0 hit
    > 2469:     opd->tls_idx = tls_next_idx++;
    dynamorio!privload_os_finalize+0x110:
    00000000`710cd120 8b05921b0c00    mov     eax,dword ptr [dynamorio!tls_next_idx (00000000`7118ecb8)] ds:00000000`7118ecb8=00000000

The call stack:

    1:004> k
     # Child-SP          RetAddr           Call Site
    00 000000ea`98d1f060 00000000`71044247 dynamorio!privload_os_finalize+0x110 [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 2469] 
    01 000000ea`98d1f0d0 00000000`71044128 dynamorio!privload_load_process+0x37 [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 835] 
    02 000000ea`98d1f100 00000000`710ccc29 dynamorio!privload_load+0x108 [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 695] 
    03 000000ea`98d1f140 00000000`710cce0e dynamorio!privload_locate_and_load+0x129 [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1874] 
    04 000000ea`98d1f1b0 00000000`710cd2aa dynamorio!privload_lookup_locate_and_load+0xfe [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1142] 
    05 000000ea`98d1f1f0 00000000`7104422e dynamorio!privload_process_imports+0xaa [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1220] 
    06 000000ea`98d1f250 00000000`71044128 dynamorio!privload_load_process+0x1e [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 830] 
    07 000000ea`98d1f280 00000000`710ccc40 dynamorio!privload_load+0x108 [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 695] 
    08 000000ea`98d1f2c0 00000000`710cce0e dynamorio!privload_locate_and_load+0x140 [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1862] 
    09 000000ea`98d1f330 00000000`710cd2aa dynamorio!privload_lookup_locate_and_load+0xfe [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1142] 
    0a 000000ea`98d1f370 00000000`710445ae dynamorio!privload_process_imports+0xaa [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 1220] 
    0b 000000ea`98d1f3d0 00000000`71043a86 dynamorio!privload_process_early_mods+0x8e [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 141] 
    0c 000000ea`98d1f550 00000000`71005866 dynamorio!loader_init_prologue+0x86 [c:\users\myself\projects\dynamorio\core\loader_shared.c @ 198] 
    0d 000000ea`98d1f590 00000000`7109c06c dynamorio!dynamorio_app_init+0x166 [c:\users\myself\projects\dynamorio\core\dynamo.c @ 569] 
    0e 000000ea`98d1f600 00000000`710d3730 dynamorio!auto_setup+0x1c [c:\users\myself\projects\dynamorio\core\arch\x86_code.c @ 169] 
    0f 000000ea`98d1f640 00007ff8`ca9ca310 dynamorio!dynamo_auto_start+0x10 [C:\Users\myself\projects\dynamorio\builds\build-vs13-7.91.18292-nodebug\core\x86.asm_core.s @ 1600] 
    10 000000ea`98d1f648 00000000`710d3720 KERNEL32!GetProcAddressStub
    11 000000ea`98d1f650 00000000`00000000 dynamorio!dynamo_auto_start [C:\Users\myself\projects\dynamorio\builds\build-vs13-7.91.18292-nodebug\core\x86.asm_core.s @ 1592] 

The variables:

    1:004> dx modpath
    modpath                 : "C:\Windows/system32/msvcrt.dll" [Type: char [260]]
    1:004> dx tls_next_idx
    tls_next_idx     : 0 [Type: int]
    1:004> dx tls_array_count
    tls_array_count  : 0 [Type: int]
    1:004> dx mod
    mod                 : 0x256ba9474b0 [Type: _privmod_t *]
        [+0x000] base             : 0x0 [Type: unsigned char *]
        [+0x008] size             : 0x0 [Type: unsigned __int64]
        [+0x010] name             : 0x256ba8f0390 : "H???" [Type: char *]
        [+0x018] path             : "P???" [Type: char [260]]
        [+0x11c] ref_count        : 0x2460 [Type: unsigned int]
        [+0x120] externally_loaded : 2 [Type: char]
        [+0x121] is_client        : 80 'P' [Type: char]
        [+0x122] called_proc_entry : 37 '%' [Type: char]
        [+0x123] called_proc_exit : 0 [Type: char]
        [+0x128] next             : 0x200002720020000 [Type: _privmod_t *]
        [+0x130] prev             : 0x28e00200002760 [Type: _privmod_t *]
        [+0x138] os_privmod_data  : 0x5002000029200200 [Type: void *]

I can’t interpret the contents of `mod`, but I verified that `modpath` has last been set in `privload_locate_and_load` in `.frame 3` of the stack, i.e., at

    /* 2) cur dir: we do not support */
    if (systemroot[0] != '\0') {
        /* 3) system dir */
        snprintf(modpath, BUFFER_SIZE_ELEMENTS(modpath), "%s/system32/%s", systemroot,
                 impname);

and that `modpath` becomes `filename` in `privload_load` in `.frame 2`. The breakpoint is never hit again before the crash:

    1:004> g
    (51f0.3a6c): Access violation - code c0000005 (first chance)
    First chance exceptions are reported before any exception handling.
    This exception may be expected and handled.
    >  385:             GLOBAL_DCONTEXT, byte *, tls_array_count, ACCT_OTHER, PROTECTED);
    dynamorio!os_loader_thread_init_prologue+0x2b:
    00000000`710cb79b 89051b350c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118ecbc)],eax ds:00000000`7118ecbc=00000001

**Comment by derekbruening** [comment]
Thank you for the detailed information.

So this is during DR's process initialization, and it loads the direct static import dbghelp.dll, which imports from msvcrt.dll, which has static TLS.  So it should increment `tls_next_idx`.  Then, when the process init is pretty much done, it should run the thread init for the main thread.  That thread init should set `tls_array_count` while `.data` is writable.  Then `.data` is made read-only, initialization is all done, and DR starts executing the main thread.  Later, this 2nd thread is created and DR initializes it, but `tls_array_count` is already set so it should not have any problems.

So I'm confused as to why that timeline is not followed here?  This is why all my earlier theories were about the 2nd thread racing with the main thread.  Do other threads exist at the time DR's process init is executing?  When that 2nd thread is initializing, the main thread is already all initialized and running under DR control, right?  Your callstack showed it *not* running DR init code, so presumably that was the case.  I would have expected its `rip` to be in the code cache though which is one suspicious datapoint.  Is `dynamo_initialized` true at that point?  Looking back: you have the main thread as thread number 4 and the other as 8 -- that seems odd to hit a problem like this that late?  Why didn't all those other threads hit it?

**Comment by debach** [comment]
Just a brief partial answer before I have to get back to work.

At the time `dynamo_initialized` is set to true in `dynamorio_app_init`, I have this situation:

    1:004> ~*k

    .  4  Id: 4ef4.ca4 Suspend: 1 Teb: 0000005d`f487b000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4a7f660 00000000`7109c06c dynamorio!dynamorio_app_init+0x321 [c:\users\myself\projects\dynamorio\core\dynamo.c @ 730] 
    01 0000005d`f4a7f6d0 00000000`710d3730 dynamorio!auto_setup+0x1c [c:\users\myself\projects\dynamorio\core\arch\x86_code.c @ 169] 
    02 0000005d`f4a7f710 00007ff8`ca9ca310 dynamorio!dynamo_auto_start+0x10 [C:\Users\myself\projects\dynamorio\builds\build-vs13-7.91.18292-nodebug\core\x86.asm_core.s @ 1600] 
    03 0000005d`f4a7f718 00000000`710d3720 KERNEL32!GetProcAddressStub
    04 0000005d`f4a7f720 00000000`00000000 dynamorio!dynamo_auto_start [C:\Users\myself\projects\dynamorio\builds\build-vs13-7.91.18292-nodebug\core\x86.asm_core.s @ 1592] 

       5  Id: 4ef4.1330 Suspend: 1 Teb: 0000005d`f487d000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4aff458 00007ff8`cad34060 ntdll!NtWaitForWorkViaWorkerFactory+0x14
    01 0000005d`f4aff460 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4aff820 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4aff850 00000000`00000000 ntdll!RtlUserThreadStart+0x21

       6  Id: 4ef4.18d0 Suspend: 1 Teb: 0000005d`f487f000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4b7f4b8 00007ff8`cad34060 ntdll!NtWaitForWorkViaWorkerFactory+0x14
    01 0000005d`f4b7f4c0 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4b7f880 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4b7f8b0 00000000`00000000 ntdll!RtlUserThreadStart+0x21

       7  Id: 4ef4.4ae8 Suspend: 1 Teb: 0000005d`f4881000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4bff788 00007ff8`cad34060 ntdll!NtWaitForWorkViaWorkerFactory+0x14
    01 0000005d`f4bff790 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4bffb50 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4bffb80 00000000`00000000 ntdll!RtlUserThreadStart+0x21

At some later point, that thread 8 is created and causes the crash. At that point, the stacks of threads 5, 6, 7 look like this (for the other threads’ traces, see above):

       5  Id: 4ef4.1330 Suspend: 1 Teb: 0000005d`f487d000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4aff458 00007ff8`cad34060 dynamorio!interception_code_array+0x1432
    01 0000005d`f4aff460 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4aff820 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4aff850 00000000`00000000 ntdll!RtlUserThreadStart+0x21

       6  Id: 4ef4.18d0 Suspend: 1 Teb: 0000005d`f487f000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4b7f4b8 00007ff8`cad34060 dynamorio!interception_code_array+0x1432
    01 0000005d`f4b7f4c0 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4b7f880 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4b7f8b0 00000000`00000000 ntdll!RtlUserThreadStart+0x21

       7  Id: 4ef4.4ae8 Suspend: 1 Teb: 0000005d`f4881000 Unfrozen
     # Child-SP          RetAddr           Call Site
    00 0000005d`f4bff788 00007ff8`cad34060 dynamorio!interception_code_array+0x1432
    01 0000005d`f4bff790 00007ff8`ca9c7bd4 ntdll!TppWorkerThread+0x300
    02 0000005d`f4bffb50 00007ff8`cad6ced1 KERNEL32!BaseThreadInitThunk+0x14
    03 0000005d`f4bffb80 00000000`00000000 ntdll!RtlUserThreadStart+0x21

Thread 4 starts at `notepad!WinMainCRTStartup`, while the other threads all start at `ntdll!TppWorkerThread`.

I will try to find out more tomorrow, or maybe starting from Tuesday.

**Comment by derekbruening** [comment]
So there are a ton of early threads floating around (in the old days this was not allowed by MS's docs), but they stack up waiting for DR init to finish as they should.  And the crashing thread comes in later.  So the mystery is: why didn't the main thread set `tls_array_count` at the end of the main process initialization when the main thread's initialization code is run and reaches `os_loader_thread_init_prologue`?

**Comment by debach** [comment]
A stupid question: might there be an erroneous optimization? The disassembly of `os_loader_thread_init_prologue` on my machine looks like:

    dynamorio!os_loader_thread_init_prologue [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 374]:
      374 00000000`710cb770 48895c2408      mov     qword ptr [rsp+8],rbx
      374 00000000`710cb775 57              push    rdi
      374 00000000`710cb776 4883ec20        sub     rsp,20h
      374 00000000`710cb77a 488bd9          mov     rbx,rcx
      380 00000000`710cb77d 8b0d35350c00    mov     ecx,dword ptr [dynamorio!tls_next_idx (00000000`7118ecb8)]
      380 00000000`710cb783 33ff            xor     edi,edi
      380 00000000`710cb785 85c9            test    ecx,ecx
      380 00000000`710cb787 7e4f            jle     dynamorio!os_loader_thread_init_prologue+0x68 (00000000`710cb7d8)  Branch

The jump corresponds to the `else` branch, which is `dcontext->priv_static_tls = NULL;`. Next comes the `if` branch:

    dynamorio!os_loader_thread_init_prologue+0x19 [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 381]:
      381 00000000`710cb789 8b052d350c00    mov     eax,dword ptr [dynamorio!tls_array_count (00000000`7118ecbc)]
      381 00000000`710cb78f 85c0            test    eax,eax
      381 00000000`710cb791 0f44c1          cmove   eax,ecx
      385 00000000`710cb794 4883c9ff        or      rcx,0FFFFFFFFFFFFFFFFh
      385 00000000`710cb798 4863d0          movsxd  rdx,eax
      385 00000000`710cb79b 89051b350c00    mov     dword ptr [dynamorio!tls_array_count (00000000`7118ecbc)],eax
      385 00000000`710cb7a1 48c1e203        shl     rdx,3
      385 00000000`710cb7a5 e816e8f5ff      call    dynamorio!heap_alloc (00000000`71029fc0)
      386 00000000`710cb7aa 4c63050b350c00  movsxd  r8,dword ptr [dynamorio!tls_array_count (00000000`7118ecbc)]
      386 00000000`710cb7b1 33d2            xor     edx,edx
      386 00000000`710cb7b3 49c1e003        shl     r8,3
      386 00000000`710cb7b7 488bc8          mov     rcx,rax
      386 00000000`710cb7ba 48898310030000  mov     qword ptr [rbx+310h],rax
      386 00000000`710cb7c1 e8368a0000      call    dynamorio!memset (00000000`710d41fc)
      387 00000000`710cb7c6 488b8310030000  mov     rax,qword ptr [rbx+310h]
      387 00000000`710cb7cd 654889042558000000 mov   qword ptr gs:[58h],rax
      390 00000000`710cb7d6 eb07            jmp     dynamorio!os_loader_thread_init_prologue+0x6f (00000000`710cb7df)  Branch

If I’m not mistaken, the write to `tls_array_count` always happens when `tls_next_idx > 0`. This seems incorrect to me.

This is the `else` branch:

    dynamorio!os_loader_thread_init_prologue+0x68 [c:\users\myself\projects\dynamorio\core\win32\loader.c @ 391]:
      391 00000000`710cb7d8 4889bb10030000  mov     qword ptr [rbx+310h],rdi

**Comment by derekbruening** [comment]
There are some synthetic static TLS client lib tests -- should check why they do not hit this.  Possibly they are used w/ just a single-threaded hello,world app.

**Comment by derekbruening** [comment]
Actually the client.tls test does have threads in the app, and it does repro this in release build locally.  I think it is not run on the CI b/c the full test time is so long there is a shrunken set run in release (again focusing on debug).
